File: rewrite-javascript/rewrite/src/rpc/request/prebuild.ts
/*
 * Copyright 2025 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as rpc from "vscode-jsonrpc/node";
import * as path from "path";
import {withMetrics} from "./metrics";
import type {ProjectDescriptor} from "../../javascript/workspace-discovery";

/**
 * Result of the Prebuild RPC: the independent projects discovered in the repository
 * (workspace-aware), each with its source sets, config-input watch-set, and parser
 * settings. The JavaScript analog of "invoke the build tool" — it produces the build
 * descriptor a consumer uses to drive builds; it does not parse sources.
 *
 * v1 is read-only: no dependency install, and {@link ProjectDescriptor.resolution} is
 * always null.
 */
export interface PrebuildResult {
    projects: ProjectDescriptor[];
}

/**
 * RPC request to compute the build descriptor for a repository (or partition) root.
 *
 * Mirrors {@link ParseProject}'s shape: optional {@code exclusions}, a {@code relativeTo}
 * base for returned paths, and a forward-compat {@code options} carrier (unused in v1).
 */
export class Prebuild {
    constructor(
        private readonly repositoryRoot: string,
        private readonly exclusions?: string[],
        /**
         * Optional base for all returned paths. Defaults to {@code repositoryRoot}.
         * Use when prebuilding a subdirectory but wanting paths relative to the repo root.
         */
        private readonly relativeTo?: string,
        /**
         * Forward-compat carrier for future options (e.g. stateful-session hints).
         * Unused in v1.
         */
        private readonly options?: Record<string, unknown>
    ) {}

    static handle(connection: rpc.MessageConnection, metricsCsv?: string): void {
        connection.onRequest(
            new rpc.RequestType<Prebuild, PrebuildResult, Error>("Prebuild"),
            withMetrics<Prebuild, PrebuildResult>(
                "Prebuild",
                metricsCsv,
                (context) => async (request) => {
                    context.target = request.repositoryRoot;

                    // Dynamic import to break circular dependency (cf. parse-project.ts).
                    const {discoverProjects} = await import("../../javascript/index.js");

                    const repositoryRoot = path.resolve(request.repositoryRoot);
                    const relativeTo = request.relativeTo ? path.resolve(request.relativeTo) : repositoryRoot;

                    const projects = await discoverProjects(repositoryRoot, {
                        exclusions: request.exclusions,
                        relativeTo
                    });

                    return {projects};
                }
            )
        );
    }
}

File: rewrite-javascript/src/main/java/org/openrewrite/javascript/rpc/Prebuild.java
/*
 * Copyright 2025 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.javascript.rpc;

import lombok.Value;
import org.jspecify.annotations.Nullable;
import org.openrewrite.rpc.request.RpcRequest;

import java.nio.file.Path;
import java.util.List;
import java.util.Map;

/**
 * RPC request to compute the build descriptor for a repository (or partition) root:
 * the independent projects (workspace-aware), their source sets, config-input watch-set,
 * and parser settings. The JavaScript analog of "invoke the build tool".
 * <p>
 * v1 is read-only — no dependency install.
 */
@Value
class Prebuild implements RpcRequest {
    /**
     * Path to the repository (or partition) root to analyze.
     */
    Path repositoryRoot;

    /**
     * Optional glob patterns to exclude from discovery.
     * If not provided, default exclusions (node_modules, dist, etc.) will be used.
     */
    @Nullable
    List<String> exclusions;

    /**
     * Optional base for all returned paths. If not specified, paths are relative to
     * {@link #repositoryRoot}. Use when analyzing a subdirectory but wanting paths
     * relative to the repository root.
     */
    @Nullable
    Path relativeTo;

    /**
     * Forward-compat carrier for future options (e.g. stateful-session hints). Unused in v1.
     */
    @Nullable
    Map<String, Object> options;

    Prebuild(Path repositoryRoot) {
        this(repositoryRoot, null, null, null);
    }

    Prebuild(Path repositoryRoot, @Nullable List<String> exclusions) {
        this(repositoryRoot, exclusions, null, null);
    }

    Prebuild(Path repositoryRoot, @Nullable List<String> exclusions, @Nullable Path relativeTo) {
        this(repositoryRoot, exclusions, relativeTo, null);
    }

    Prebuild(Path repositoryRoot, @Nullable List<String> exclusions, @Nullable Path relativeTo, @Nullable Map<String, Object> options) {
        this.repositoryRoot = repositoryRoot;
        this.exclusions = exclusions;
        this.relativeTo = relativeTo;
        this.options = options;
    }
}
